### PR TITLE
[docs] Make spelling of 'nonnegative' consistent in TripletMarginWithDistanceLoss

### LINKED ISSUE #43342: Proposal: Generic Triplet-Margin Loss
## 🚀 Feature
We'd like to add a distance-agnostic version of the [TripletMarginLoss](https://pytorch.org/docs/master/generated/torch.nn.TripletMarginLoss.html), as the current version only supports l_p norms.  This was something we implemented as part of a recent investigation for embedding learning and thought could be helpful to the PyTorch userbase at large.

## Motivation

1. Extensibility: decoupling the distance function and the loss computation allow for more flexibility for users.  PyTorch currently has a [CosineEmbeddingLoss](https://pytorch.org/docs/stable/generated/torch.nn.CosineEmbeddingLoss.html), but that serves a somewhat different purpose and doesn't really work for users wanting a triplet-margin loss with cosine distance.

2. Existing use cases: several papers have proposed triplet loss functions with cosine distance ([1](https://www.isca-speech.org/archive/Interspeech_2018/pdfs/1209.pdf), [2](https://www.researchgate.net/publication/316736728_Deep_Speaker_an_End-to-End_Neural_Speaker_Embedding_System#read)) or have generally used cosine-based metrics ([1](https://openaccess.thecvf.com/content_cvpr_2018/papers/Wang_CosFace_Large_Margin_CVPR_2018_paper.pdf), [2](https://arxiv.org/pdf/1812.00442.pdf)).  [PyTorch-BigGraph](https://arxiv.org/abs/1903.12287) also does something similar with its ranking loss.

3.  Products like Tensorflow [decouple](https://github.com/tensorflow/addons/blob/v0.10.0/tensorflow_addons/losses/triplet.py) the distance functions and even allow for custom distance metrics

## Pitch

We have a working implementation of this (see screenshots) and propose to use it directly, with adjustments to fit the existing PyTorch code structure.  We'll tentatively call these `torch.nn.GenericTripletMarginLoss` and `F.generic_triplet_margin_loss`.  In the proposed implementation, `GenericTripletMarginLoss` will be a stateful wrapper for `F.generic_triplet_margin_loss`, and the loss computation will occur in the latter.

The module signatures will resemble that of the [production module](https://pytorch.org/docs/master/generated/torch.nn.TripletMarginLoss.html), with an analogous signature for the [functional module](https://pytorch.org/docs/master/_modules/torch/nn/modules/loss.html#TripletMarginLoss).  The removed parameters (`p` and `eps`) are specific to l_p norm computation; `size_average` and `reduce` are removed because of deprecation.  There are two new parameters: `distance_function` and `is_similarity_function`.  `distance_function` is a function/module $$f: R^{n \cross d} \rightarrow R^n$$ that gives an index-wise distance metric between two tensors of embeddings.  This can be, for example, `torch.nn.PairwiseDistance` or `torch.nn.CosineSimilarity`.  `is_similarity_function` is a boolean that denotes whether `distance_function` is a distance or similarity function: this will indicate whether to flip the signs in distance computation.

## Alternatives

Beyond the straightforward alternative of leaving this to the user to implement every time, we haven't come up with any other alternatives.

One open question is whether the Python implementation is performant enough, or if this would need to be added in aTen as well.

## Additional context

image

cc @albanD @mruberry

### PR DESCRIPTION
**Pull request by ethch18**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45378 [nn] Make spelling of 'nonnegative' consistent in TripletMarginWithDistanceLoss**

This PR fixes some inconsistencies in the docs of `nn.TripletMarginWithDistanceLoss`, specifically w.r.t. the spelling of the word "nonnegative" (vs. "non-negative").  It seems that "nonnegative" is more common in general, but I'm not sure if there's a preference; I've changed all instances to that for now but am happy to go the other way.

nonnegative

### CODE REVIEW

**Comment by mruberry** [comment]
Seems fine. If you like you can include this fix with another more substantial PR. I'd prefer we didn't spin up a bunch of internal builds just for this.

**Comment by ethch18** [comment]
Sure, does it have to be a Python PR?  I have PR #45377 open but it's a C++ one, not sure if that will clutter things more.

**Comment by mruberry** [comment]
> Sure, does it have to be a Python PR? I have PR #45377 open but it's a C++ one, not sure if that will clutter things more.

Adding it to #45377 sounds good.

**Comment by ethch18** [comment]
Closing this since I've incorporated it into #45377
